src/hotspot/share/classfile/bytecodeAssembler.cpp

```diff
@@ -25,33 +25,41 @@
 #include "precompiled.hpp"
 
 #include "classfile/bytecodeAssembler.hpp"
+#include "classfile/vmSymbols.hpp"
 #include "interpreter/bytecodes.hpp"
 #include "memory/oopFactory.hpp"
 #include "oops/constantPool.hpp"
 #include "runtime/handles.inline.hpp"
 #include "utilities/bytes.hpp"
 
-u2 BytecodeConstantPool::find_or_add(BytecodeCPEntry const& bcpe) {
+u2 BytecodeConstantPool::find_or_add(BytecodeCPEntry const& bcpe, TRAPS) {
 
-  u2 index = _entries.length();
+  // Check for overflow
+  int new_size = _orig->length() + _entries.length();
+  if (new_size > USHRT_MAX) {
+    THROW_MSG_0(vmSymbols::java_lang_InternalError(), "default methods constant pool overflowed");
+  }
+
+  u2 index = static_cast<u2>(_entries.length());
   bool created = false;
   u2* probe = _indices.put_if_absent(bcpe, index, &created);
   if (created) {
     _entries.append(bcpe);
   } else {
     index = *probe;
   }
-  return index + _orig->length();
+  return static_cast<u2>(index + _orig->length());
```

Comment:
The static_cast<u2> should be changed to checked_cast<u2> since we know that we will never truncate.

Reply:
If we know they're always in range, then why check?  Changed them.